Creation of and switch to using generic DataStorage object for managing data

KillerOfPie opened the pull request:
[contributing]: https://github.com/SparklingComet/TradeShop/blob/master/.github/CONTRIBUTING.md

This PR is mainly for: https://github.com/Tradeshop/TradeShop/projects/2#card-33848153

## Pull Request Etiquette

There are several guidelines you should follow in order for your
  Pull Request to be merged.

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

> It is sometimes better to include more changes in a single commit. 
  If you find yourself having an overwhelming amount of commits, you
  can **rebase** your branch.

## Template

### Pull Request

Replace this sentence with general description of what your Pull Request does.

KillerOfPie committed:
Creation of and switch to using generic DataStorage object for managing data

KillerOfPie committed:
Creation of and switch to using generic DataStorage object for managing data

KillerOfPie commented:
I did some basic testing on this and it seems to be functioning as is

SparklingComet commented:
I think we are going to wait a moment until 2.2.2 is out and make sure there are no bugs reported about this. No harm doing preparations im advance, but SQL-based data storage will wait until 3.0

KillerOfPie committed:
Changes and new data storage
 - Changes to make player data storage more versatile
 - Added relevant shops to player data file

KillerOfPie committed:
Changes and new data storage
 - Changes to make player data storage more versatile
 - Added relevant shops to player data file

KillerOfPie committed:
Catch for bad data setting
 - Changes to catch and stop plugin when data-storage-type is improperly set
 - Changes to Debug system to allow negative positions for error codes

KillerOfPie committed:
Catch for bad data setting
 - Changes to catch and stop plugin when data-storage-type is improperly set
 - Changes to Debug system to allow negative positions for error codes

KillerOfPie committed:
Added stored Chest Linkage data for Json
 - May remove persistent data values later if this works well
 - Also fixed a missing .getString for a setting in Utils

KillerOfPie committed:
Added stored Chest Linkage data for Json
 - May remove persistent data values later if this works well
 - Also fixed a missing .getString for a setting in Utils

KillerOfPie commented:
K, I think I have this pretty well flushed out, let me know if this needs any changes or any addition features added for the generic data object. Otherwise I think this feature done for 2.3.0. I did basic testing and didn't find any problems, of course more thorough testing will be needed later

KillerOfPie committed:
Bug fixes from discord
- Fixed bug involving world names including '-'
- Fixed protection not detecting other half of double chests

KillerOfPie committed:
Bug fixes from discord
- Potential fix for NPE in restock listener, with currently unknown cause

KillerOfPie committed:
Bug fix for #89

KillerOfPie committed:
Potential Fix for https://discordapp.com/channels/583000924528836639/583387430305136670/763117283295363103

KillerOfPie committed:
Update Version to 2.2.3

KillerOfPie committed:
Updating ShopChunk to match ShopLocation for world name error

KillerOfPie committed:
Properly Fixed ShopChunk serialization and fixes existing data on file load

KillerOfPie committed:
Minor Bug fixes and Changes

KillerOfPie committed:
*Fix* for untraced NPE
Version changed to STABLE

SparklingComet committed:
Fix .gitignore

KillerOfPie committed:
*Fixes* + 2.2.4-STABLE increment
- Fix for unnecessary debug object creation
- Fix for 1st slot shulker error

KillerOfPie committed:
Merge pull request #90 from Tradeshop/2.2.3

2.2.3+2.2.4

KillerOfPie committed:
Merge remote-tracking branch 'origin/Data-Management' into Data-Management

# Conflicts:
#	src/main/java/org/shanerx/tradeshop/utils/data/JsonConfiguration.java

SparklingComet commented on src/main/java/org/shanerx/tradeshop/listeners/ShopProtectionListener.java:
```diff
@@ -58,43 +58,48 @@
+                debugger.log("ShopProtectionListener: HopperEvent isForbidden: " + isForbidden, DebugLevels.PROTECTION);
+            }
+        } catch (NullPointerException ignored) {
+        } // Fix for random NPE triggering from this event that shows no stack trace
```
what could trigger the NPE?

KillerOfPie replied:
No clue it was just a random NPE that popped in console during the hopper event but nothing was given. Hopper protection still worked properly so i figured this is the best fix without spending tons of time digging for it.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/listeners/ShopProtectionListener.java:
```diff
@@ -217,7 +222,7 @@ public void onChestOpen(PlayerInteractEvent e) {
         Block block = e.getClickedBlock();
 
-        if (!(e.getAction() == Action.RIGHT_CLICK_BLOCK && plugin.getListManager().isInventory(block))) {
+        if (e.getAction() != Action.RIGHT_CLICK_BLOCK || !plugin.getListManager().isInventory(block)) {
```
this is just refactoring, no change in the logic

KillerOfPie replied:
Correct, just made it easier to read

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/CommandTabCompleter.java:
```diff
@@ -71,7 +71,7 @@ public CommandTabCompleter(TradeShop instance, CommandPass command) {
 	public List<String> addSet() {
 		if (command.argsSize() == 2) {
-			return Arrays.asList("1", "2", "4", "8", "16", "32", "64", "96", "128");
+			return Arrays.asList("1", "2", "4", "8", "16", "32", "64", "80", "96", "128");
```
sorry but what does 80 mean? also why are these strings?

KillerOfPie replied:
These are tab completion values for amounts, 80 is a stack + 16. Tikkle requested it be added as it was am amount that his players used. These are strings because that's what the tab completion uses and putting them as ints to just be converted would be unnecessary and wasteful.

SparklingComet replied:
Fair enough

SparklingComet commented on src/main/java/org/shanerx/tradeshop/listeners/ShopTradeListener.java:
```diff
@@ -174,7 +174,7 @@ public void onBlockInteract(PlayerInteractEvent e) {
     private boolean tradeAllItems(Shop shop, int multiplier, Action action, Player buyer) {
         ArrayList<ItemStack> costItems = new ArrayList<>(), productItems = new ArrayList<>();
-        Inventory shopInventory = shop.getChestAsSC().getInventory();
+        Inventory shopInventory = shop.hasStorage() ? shop.getChestAsSC().getInventory() : null;
```
where is the variable `shopInventory` being used? I cannot find any further references in this function

KillerOfPie replied:
This is used when checking the inventory of the shop in the #getItems method, it can be null for iTrade shops hence the lamda if

Used on lines 214 & 236

SparklingComet commented:
wait what happened here? why was this closed?

SparklingComet commented on src/main/java/org/shanerx/tradeshop/listeners/ShopCreateListener.java:
```diff
@@ -79,9 +78,7 @@ public void onSignChange(SignChangeEvent event) {
-		JsonConfiguration chunk = new JsonConfiguration(shopSign.getChunk());
-
-		if (Setting.MAX_SHOPS_PER_CHUNK.getInt() <= chunk.getShopCount() + 1) {
+		if (Setting.MAX_SHOPS_PER_CHUNK.getInt() <= plugin.getDataStorage().getShopCountInChunk(shopSign.getChunk()) + 1) {
```
what's this changing?

KillerOfPie replied:
it changes to use the DataStorage object rather then having the method in ShopChunk call from JsonConfiguration

SparklingComet commented on src/main/java/org/shanerx/tradeshop/utils/Utils.java:
```diff
@@ -62,7 +63,7 @@
 	private final UUID KOPUUID = UUID.fromString("daf79be7-bc1d-47d3-9896-f97b8d4cea7d");
 	private final UUID LORIUUID = UUID.fromString("e296bc43-2972-4111-9843-48fc32302fd4");
-	protected TradeShop plugin = (TradeShop) Bukkit.getPluginManager().getPlugin("TradeShop");
+	public TradeShop plugin = (TradeShop) Bukkit.getPluginManager().getPlugin("TradeShop");
```
maybe make `final`?

KillerOfPie replied:
yes, don't know why it wasnt.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/utils/data/DataType.java:
```diff
@@ -0,0 +1,31 @@
+public enum DataType {
+    FLATFILE,
+    SQLITE
+}
```
ok.

KillerOfPie replied:
I didn't add in support for SQLite but i did need more than 1 value in the enum so I just dropped that in.

SparklingComet reviewed (commented):
k saw you closed this and moved to #92 so I am dumping the review
